### PR TITLE
multer-s3 - extended req and file definition to work with @koa/multer types

### PR DESCRIPTION
**Pull request by abelkius**
I extended types in this definition to also work with `koa-multer` package. The only difference is the type for `req` that is passed through and `S3Storage` type that also differs just because of the `req` that is passed through. 

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/koa-multer/index.d.ts
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`. If for reason the any rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]` and not for whole package so that the need for disabling can be reviewed.

### CODE REVIEW

**Review by andrewbranch — CHANGES_REQUESTED** [review]
Hm, I think I see the problem you’re trying to solve, but this isn’t the right approach. Really, the type of `req` should be generic. While the status quo is wrong, and this PR is arguably a step in the right direction, it would cause breakages for people who _are_ using Express and relying on the types to be express-like.

Unfortunately I'm not familiar enough with these packages to be able to tell you exactly what this would look like. If you can post an example of how this is used with Koa, that would be helpful.

**Comment by abelkius** [comment]
Hi @andrewbranch, thank you a ton for your review. I updated the PR and now I use generics in the exported interface. The only problem that appeared with this approach is that I needed to disable the `no-unnecessary-generics` in few places. I hope it is not a problem, because I don't think I should actually use `unknown` type in there. What do you think?

File: types/multer-s3/index.d.ts [hunk 1]
```diff
@@ -3,54 +3,60 @@
...
-        req: Express.Request,
-        file: Express.Multer.File,
-        callback: (error: any, mime?: string, stream?: NodeJS.ReadableStream) => void): void;
-    DEFAULT_CONTENT_TYPE(
-        req: Express.Request,
-        file: Express.Multer.File,
-        callback: (error: any, mime?: string) => void): void;
+    interface S3Storage {
+        // tslint:disable-next-line:no-unnecessary-generics
+        <E, R>(options?: Options<R>): E;
```
Comment (weswigham) [inline types/multer-s3/index.d.ts:47]: Why is the return type not still `StorageEngine`? I don't think the caller can change the fact that a `S3Storage` instance is returned.

File: types/multer-s3/index.d.ts [hunk 2]
```diff
@@ -3,54 +3,60 @@
...
+        /** Location of the uploaded file (DiskStorage) */
+        path: string;
+        /** A Buffer of the entire file (MemoryStorage) */
+        buffer: Buffer;
     }
-}
 
-interface S3Storage {
-    (options?: Options): StorageEngine;
+    interface Options<R> {
```
Comment (weswigham) [inline types/multer-s3/index.d.ts:34]: Consider adding `R = Express.Request` for backwards compatibility.

File: types/multer-s3/index.d.ts [hunk 3]
```diff
@@ -3,54 +3,60 @@
...
-        file: Express.Multer.File,
-        callback: (error: any, mime?: string, stream?: NodeJS.ReadableStream) => void): void;
-    DEFAULT_CONTENT_TYPE(
-        req: Express.Request,
-        file: Express.Multer.File,
-        callback: (error: any, mime?: string) => void): void;
+    interface S3Storage {
+        // tslint:disable-next-line:no-unnecessary-generics
+        <E, R>(options?: Options<R>): E;
+        AUTO_CONTENT_TYPE<R>(
```
Comment (weswigham) [inline types/multer-s3/index.d.ts:48]: In these cases, you can actually say that `req` is `any`. The implementation quite literally does not inspect the parameter, and does not care a whit about its value. That should make it so you don't need the type parameters here.

**Review by weswigham — CHANGES_REQUESTED** [review]

**Pull request closed without merging**